### ISSUE TITLE
cdk_route53: cdk.context.json causes HostedZone lookup to fail

### ISSUE THREAD

**Issue by adonig** [aws/aws-cdk#23437]
### Describe the bug

I ran into a 404 error after manually deleting a hosted zone created by CDK, recreating the zone with the AWS console and then trying to look up the zone with the CDK by name. The lookup fails with a 404 trying to find a zone with a wrong ID.

### Expected Behavior

The CDK should be able to find the zone.

### Current Behavior

No hosted zone found with ID: Z049836444COOXK7EL2O (Service: AmazonRoute53; Status Code: 404; Error Code: NoSuchHostedZone; Request ID: fffc8a5a-8e70-4c40-b83b-e1d7946735e8; Proxy: null)

### Reproduction Steps

1. Create a hosted zone with CDK.
2. Delete it manually and recreate it manually in the AWS Console.
3. Lookup the zone with CDK by name.

### Possible Solution

I used grep and found the old ID in a file called `cdk.context.json`. After deleting the file I was able to deploy my stack again. The file got recreated and now contains the current ID of the hosted zone.

What's the reason to write the HostedZone into this file? If it's caching, maybe it can get cached somewhere else, i.e. `cdk.out/.cache` or something like that? The CDK should also repeat the lookup without the cache in case of a 404.

If all that isn't possible, maybe the CDK should point you in case of a failed lookup to a troubleshooting guide.

### Additional Information/Context

Deleting `cdk.out` and `__pycache__` doesn't solve the problem.

### CDK CLI Version

2.56.0 (build 1485f48)

### Node.js Version

v18.12.1

### OS

Fedora Linux 37

### Language

Python

### Language Version

3.11.0

### Other information

Here is why I deleted and recreated the hosted zone manually with the AWS console: I first created a hosted zone for my TLD with CDK. Then I realized that it's better to have a [hosted zone for the subdomain](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/dns-routing-traffic-for-subdomains.html) in my use-case. So I deleted the one created by the CDK and created a new one using the AWS console. Then I changed my CDK template to create the zone for my subdomain and lookup the one for the TLD to add a delegation NS record there.

**Comment by peterwoodworth** [comment]
Be sure to read our [guide on runtime context](https://docs.aws.amazon.com/cdk/v2/guide/context.html) to fully understand why we do things this way, and how to manage context. This should clear up all the questions you have, let me know if now

**Comment by adonig** [comment]
Hi @peterwoodworth! Thanks for the reference. I understand that a context is necessary, but what's the benefit of having the lookup by name fail with a 404 instead of returning the actual hosted zone?

**Comment by peterwoodworth** [comment]
From my understanding reading your issue, is that the previous lookup was cached within `cdk.context.json`. Thus, the lookup for your new hosted zone never occurred because the previous value was cached. The deployment went ahead using your old cached values, and deployment failed because these values defined in your template were outdated. Is this correct?

If the above is correct, then the lookup isn't what failed and gave a 404 - it was the subsequent CloudFormation deployment

**Comment by adonig** [comment]
Yes, that's possible. It looks like it failed when it tried to use a stale reference to a hosted zone that doesn't exist. It would be much better if we could prevent that.

Can't we mark the hosted zone as "from_cache" and when it fails we can ask the user whether he wants to delete the stale reference from the cache and run it again?

**Comment by peterwoodworth** [comment]
That's a good idea for a feature request! I'm not sure exactly how we can achieve that, but it would be great to let the user know if any values in `cdk.context.json` is outdated. Thanks for the suggestion 🙂

**Comment by adonig** [comment]
It's the principle of least surprise. The user expects the CDK to deploy a stack and is surprised when it fails referencing a non-existing resource.

Nevertheless the CDK is a great software and makes life so much better than when we had to write CloudFormation templates by hand. I tried to build [something alike](https://github.com/adonig/aloisius) in 2017 and I'm kind of happy I didn't succeed because otherwise I would have to maintain it.

So let me say thank you for making all this happen and merry Christmas.

**Comment by willrust** [comment]
Using 'fromLookup' is where this often burns me. Some way of ignoring cache on fromLookup would be helpful.
